src/node_url.cc

```diff
@@ -1351,6 +1351,53 @@ namespace url {
                             v8::NewStringType::kNormal).ToLocalChecked());
   }
 
+  static void ToUSVString(const FunctionCallbackInfo<Value>& args) {
+    Environment* env = Environment::GetCurrent(args);
+    CHECK_GE(args.Length(), 1);
+    Local<String> str;
+    if (!args[0]->ToString(env->context()).ToLocal(&str))
+      return;
+
+    // The algorithm below can be replaced with `Utf8Value`, which
+    // automatically replaces invalid Unicode surrogates with U+FFFD
+    // REPLACEMENT CHARACTER as the Web IDL spec mandates through V8's
+    // `String::REPLACE_INVALID_UTF8`. However, using `Utf8Value` would prevent
+    // us from using the `changed` optimization, and V8 would have to create a
+    // new string even when it is not needed.
+
+    TwoByteValue value(env->isolate(), str);
+    const size_t n = value.length();
+    bool changed = false;
+
+    for (size_t i = 0; i < n; i++) {
+      uint16_t c = value[i];
+      if (c < 0xD800 || c > 0xDFFF) {
+        continue;
+      } else if (0xDC00 <= c || i == n - 1) {
+        changed = true;
+        value[i] = 0xFFFD;
+      } else {
+        uint16_t d = value[i + 1];
+        if (0xDC00 <= d && d <= 0xDFFF) {
+          i++;
+        } else {
+          changed = true;
+          value[i] = 0xFFFD;
+        }
+      }
+    }
```

Comment:
You can probably get a bit of speedup if you split this loop in a simple linear scan for invalid characters, and a second one that fixes them up.

That optimizes for the common case that of valid UTF-8 and because you can do the scan in JS land with a regex (`/[\uD800-\uDBFF]?[\uDC00-\uDFFF]/` if I read the logic correctly), you can avoid a call into C++ land.

Doing the scan in JS land penalizes the uncommon case because C++ lands needs to scan again.  Whether the first scan should be in done in JS or C++ land depends on how common the uncommon case is.  Not very, I expect.